include/ada/url_aggregator-inl.h

```diff
@@ -139,60 +137,64 @@ inline void url_aggregator::update_base_search(std::string_view input) {
     return;
   }
 
-  // Make sure search is deleted and hash_start index is correct.
-  if (components.search_start != url_components::omitted) { // Uncommon path
-    if (components.hash_start != url_components::omitted) {
-      buffer.erase(components.search_start, components.hash_start - components.search_start);
-      components.hash_start = components.search_start;
+  if (input[0] == '?') { input.remove_prefix(1); }
+
+  if (components.hash_start == url_components::omitted) {
+    if (components.search_start == url_components::omitted) {
+      components.search_start = uint32_t(buffer.size());
+      buffer += "?";
     } else {
-      buffer.erase(components.search_start);
+      buffer.resize(components.search_start + 1);
     }
-  }
 
-  uint32_t input_size = uint32_t(input.size());
-  components.search_start = components.pathname_start + get_pathname_length();
-  // The common case here is components.search_start == buffer.size().
-
-  if (input[0] != '?') {
-    // If input does not start with "?", we need to add it.
-    buffer.insert(components.search_start, helpers::concat("?", input));
-    input_size++;
+    buffer.append(input);
   } else {
-    buffer.insert(components.search_start, input);
+    if (components.search_start == url_components::omitted) {
+      components.search_start = components.hash_start;
+    } else {
+      buffer.erase(components.search_start, components.hash_start - components.search_start);
+      components.hash_start = components.search_start;
+    }
+
+    buffer.insert(components.search_start, "?");
```

Comment:
If you use the trick from PR https://github.com/ada-url/ada/pull/300, it is likely you can create a variation on update_base_search which takes an input a view that has *already* the '?' character, so instead of two inserts, you can have just one.

Reply:
I am concerned with the amount of code duplication we have. Even optimizing it is really hard (to trace and maintain) at the moment.